### ISSUE TITLE
mmfields not parsing CEF

### ISSUE THREAD

**Issue by sbakke** [rsyslog/rsyslog#98]
Hi,

As described in http://kb.monitorware.com/mmfields-broken-t12256.html and http://bugzilla.adiscon.com/show_bug.cgi?id=515

I am using Rsyslog with mmfields to receive CEF formated events from Arcsight over syslog udp and tcp. In addition to a local configuration I have the following rsyslog configuration to receive remote CEF formated messages:

module(load="imtcp")
module(load="imudp")
module(load="mmfields")
template(name="cef" type="string" string="%$!%\n")
ruleset(name="remote") {
action(type="mmfields")
action(type="omfile" file="/var/log/remote.log")
action(type="omfile" file="/var/log/remote_ parsed.log" template="cef")
}
input(type="imtcp" port="514" ruleset="remote")
input(type="imudp" port="514" ruleset="remote")

Receiving the events and logging them to remote.log is no problem. Parsing them, however, is a completely different story:

Output of remote.log (unparsed CEF), properly received and stored as expected:
2014-02-18T11:44:37.292916+01:00 somehost.somedomain.com CEF: 0|ArcSight|Logger|5.3.1.6838.0|eps:103|Individual Forwarder EPS|1| cat=/Monitor/Forwarder/EPS/Individual cn1=0 cn1Label=EPS cs1=UDP cs1Label=Forwarder Type cs2=SinceLastMonitorEvent cs2Label=timeframe cs6=log_udp cs6Label=Forwarder name dst=10.0.0.100 dvc=10.0.0.10 end=1392716400079 rt=1392716400079

Output of remote_parsed.log (parsed CEF), obviously parsed incorrectly:
{ "f1": "0|ArcSight|Logger|5.3.1.6838.0|eps:103|Individual Forwarder EPS|1| cat=\/Monitor\/Forwarder\/EPS\/Individual cn1=0 cn1Label=EPS cs1=UDP cs1Label=Forwarder Type cs2=SinceLastMonitorEvent cs2Label=timeframe cs6=log_udp cs6Label=Forwarder name dst=10.0.0.100 dvc=10.0.0.10 end=1392716400079 rt=1392716400079" }

For sake of completeness, I have included the output of rsyslogd -N1 below:
rsyslogd: version 8.1.5, config validation run (level 1), master config /etc/rsyslog.conf
rsyslogd: End of config validation run. Bye.

Just a quick note, the original data sent from ArcSight is (according to netcat):

CEF:0|ArcSight|Logger|5.3.1.6838.0|eps:103|Individual Forwarder EPS|1| cat=/Monitor/Forwarder/EPS/Individual cn1=0 cn1Label=EPS cs1=UDP cs1Label=Forwarder Type cs2=SinceLastMonitorEvent cs2Label=timeframe cs6=log_udp cs6Label=Forwarder name dst=10.0.0.100 dvc=10.0.0.10 end=1392716400079 rt=1392716400079

**Comment by vdesabou** [comment]
Did you have any update on this? I've got the same issue. Did you figure out how to work with mmfields ?
Thanks

**Comment by rgerhards** [comment]
I am working on a parser for CEF as part of the scheduled liblognorm enhancements: https://github.com/rsyslog/liblognorm/issues/40 This will be usable via mmnormalize inside rsyslog. Now that I have looked at the format, I don't think mmfields is a good match to handle CEF.

If someone could contribute a somewhat larger set of sample CEF messages, it would definitely help me to get this done. Lack of samples is always a problem, and in order to create a decent solution I think I need at least a few thousand messages, so that I can iron out any troubles.

**Comment by vdesabou** [comment]
>I am working on a parser for CEF as part of the scheduled liblognorm enhancements: rsyslog/liblognorm#40 This will be usable via mmnormalize inside rsyslog. Now that I have looked at the format, I don't think mmfields is a good match to handle CEF.

I'm really a newbie with rsyslog, what I would like to understand is if we can (either with mmfields or the future liblognorm enhancement) do the following:

Receive a CEF message like this:

```
CEF:0|AppName|AppVendor|1.0|rule:101|TEST-UNIX RULE|High| field1=a value field3=another value field4=some other value field7=0
```

And transforms it into this (fixed number of fields, so that we can easily parse messages afterwards):

```
CEF:0|AppName|AppVendor|1.0|rule:101|TEST-UNIX RULE|High| field1=a value field2= field3=another value field4=some other value field5= field6= field7=0
```

Do you have any idea if that would be possible?

>If someone could contribute a somewhat larger set of sample CEF messages, it would definitely help me to get this done. Lack of samples is always a problem, and in order to create a decent solution I think I need at least a few thousand messages, so that I can iron out any troubles.

Can't commit on this, but I'm waiting for "anonymized" CEF logs. I will share them if I can get them.

Thanks

**Comment by rgerhards** [comment]
I can confirm that those transformations are possible, *but* I would suggest to use JSON for the transformed message, as it is much cleaner and easier to parse (the scripting language of your choice almost always already has a JSON parser, so it's usually extremely easy to work with).

What concerns me a bit is the sample format that you give. If that is what the message **actually** look like, it's probably very hard to parse them in the first place. I think the spec also said it's a more clean name=value format ... but nothing beats practice. This is why I am so insistive on my couple of thousand real world sample message (please do not include whitespace where none was during anonymization). 

If you can provide me these, I promise I'll look very quickly at the parser. I am probably as desperately looking for the samples as you are for the solution ;)

**Comment by rgerhards** [comment]
I received some test data and am working on the liblognorm implementation. Now that I have seen how  evil that CEF format is, I guess mmfields does not really do what it promises. It was written based on a subset of the information I have now available on CEF. So we probably need to change the doc. I won't upgrade mmfields to include all weirdness of CEF (most notably spaces in unquoted values).

Those interested in the CEF parser, please follow https://github.com/rsyslog/liblognorm/issues/40

**Comment by rgerhards** [comment]
This has been implemented in liblognorm v2 and is available via mmnormalize.
